juju/juju
# feat(dqlite): wire up WatchMachines for provisioner via modelmachineswatcher

This adds the following:

- wires up the new `WatchMachines` watcher from the machine domain within the modelmachineswatcher (that the provisioner, firewaller and the instancepoller uses).
- severs the connection between the firewaller api on the apiserver with the modelmachineswatcher and changes the firewaller worker to use the machine service directly.
- we currently don't have all the necessary functions in the machine domain to have the instancepoller worker to use the machine service directly for everything instead of going through the api. So this wires up the machine service into the instancepoller manifold, so the instancepoller worker is using both the api and the service right now. We'll gradually remove the connection to the apiserver and replace their calls in the worker with the machine services. Currently this does replace the `WatchModelMachines` with the new `WatchMachines` so it takes out the modelmachineswatcher from the instancepoller api on the apiserver. (created  [JUJU-6310](https://warthogs.atlassian.net/browse/JUJU-6310?atlOrigin=eyJpIjoiNjY3OGRiMjRmMTU2NGFhODhhNmVlMDc0NGEwOWRlOGUiLCJwIjoiaiJ9) with the list of functions needed).

## Checklist




## QA steps

```
$ juju bootstrap aws deleteme
$ juju deploy ubuntu
$ juju add-machine
```

Take a look at the debug-log output on the model, and you should be able to spot the following:

1. for firewaller:
``` 
controller-0: 00:43:44 INFO juju.worker.firewaller opening port ranges [22/tcp from 0.0.0.0/0,::/0] on model firewall
controller-0: 00:43:44 DEBUG juju.worker.firewaller started watching "machine-0" 
```

2. for instancepoller

```
controller-0: 00:43:48 INFO juju.worker.instancepoller machine "0" (instance ID "i-0959d77d25b31b3f7") instance status changed from {"allocating" "Start instance attempt 1"} to {"running" "running"}
controller-0: 00:43:48 INFO juju.worker.instancepoller machine "0" (instance ID "i-0959d77d25b31b3f7") has new addresses: [local-cloud:172.31.21.16@alpha public:3.80.79.242@alpha]
```

for the instancepoller you can also take a look at the `show-machine` output and see if it got assigned the public address and the status is updated (to running).

3. for provisioner (deprovisioning a machine)

```
juju remove-machine 0
```

```
controller-0: 21:45:18 INFO juju.worker.provisioner stopping known instances [juju-f202b5-0]
controller-0: 21:45:21 INFO juju.worker.provisioner removing dead machine "0"
```

#### Some relevant unit tests

```
TEST_PACKAGES="./apiserver/common/modelmachineswatcher/... -count=1 -race" make run-go-tests
```

```
TEST_PACKAGES="./apiserver/facades/agent/provisioner/... -count=1 -race" make run-go-tests
```

```
TEST_PACKAGES="./apiserver/facades/controller/firewaller/... -count=1 -race" make run-go-tests
```

```
TEST_PACKAGES="./apiserver/facades/controller/instancepoller/... -count=1 -race" make run-go-tests
```

## Links

**Jira card:** JUJU-6108



[JUJU-6310]: https://warthogs.atlassian.net/browse/JUJU-6310?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
